Lib/ufo2ft/instructionCompiler.py

```diff
@@ -0,0 +1,227 @@
+                        "no instructions in font."
+                    )
+                    continue
+
+                # Check if glyph hash matches the current outlines
+                hash_pen = HashPointPen(glyph.width, self.ufo)
+                glyph.drawPoints(hash_pen)
+                glyph_id = ttdata.get("id", None)
+                if glyph_id is None or glyph_id != hash_pen.hash:
+                    logger.error(
+                        f"Glyph hash mismatch, glyph '{name}' will have "
+                        "no instructions in font."
+                    )
+                    continue
+
+                # Compile the glyph program
+                asm = ttdata.get("assembly", None)
+                if asm is not None:
+                    glyf.program = ttLib.tables.ttProgram.Program()
+                    glyf.program.fromAssembly(asm)
+                    # Roundtrip once, or if the font is dumped to XML before
+                    # having been saved, the assembly code if will look awful.
+                    glyf.program._assemble()
+                    glyf.program._disassemble(preserve=True)
+
+            # Handle composites
+            if glyf.isComposite():
+                # Remove empty glyph programs from composite glyphs
+                if hasattr(glyf, "program") and not glyf.program:
+                    delattr(glyf, "program")
+
+                # Set component flags
+
+                # We need to decide when to set the flags automatically.
+                # Let's assume if any lib key is not there, or the component
+                # doesn't have an identifier, we should autoset all component
+                # flags.
+                for i, c in enumerate(glyf.components):
+                    if i >= len(glyph.components):
+                        logger.error(
+                            "Number of components differ between UFO and TTF "
+                            f"in glyph '{name}' ({len(glyph.components)} vs. "
+                            f"{len(glyf.components)}, not setting flags in "
+                            "additional components."
+                        )
+                        break
+                    ufo_component = glyph.components[i]
+                    if (
+                        ufo_component.identifier is None
+                        or "public.objectLibs" not in glyph.lib
+                        or "public.objectIdentifiers"
+                        not in glyph.lib["public.objectLibs"]
+                        or ufo_component.identifier
+                        not in glyph.lib["public.objectLibs"][
+                            "public.objectIdentifiers"
+                        ]
+                        or "public.truetype.instructions"
+                        not in glyph.lib["public.objectLibs"][
+                            "public.objectIdentifiers"
+                        ][ufo_component.identifier]
+                    ):
```

Comment:
This should use `glib.lib["public.objectLibs"][ufo_component.identifier]` directly.